git_theta/updates/base.py

```diff
@@ -9,40 +10,148 @@
+        repo,
+        path: str,
+    ) -> metadata.ParamMetadata:
+        """Get the metadata from the last time this parameter was updated via git."""
+        logging.debug(f"Getting previous metadata for {'/'.join(param_keys)}")
+        logging.debug(f"Current Metadata for {'/'.join(param_keys)}: {param_metadata}")
+        last_commit = param_metadata.theta_metadata.last_commit
+        # TODO: Currently, if the model checkpoint is added during the first commit
+        # then we can't do a sparse update until a second dense update is commited.
+        if not last_commit:
+            raise ValueError(
+                f"Cannot find previous version for parameter {'/'.join(param_keys)}"
+            )
+        logging.debug(
+            f"Getting metadata for {'/'.join(param_keys)} from commit {last_commit}"
+        )
+        last_metadata_obj = git_utils.get_file_version(repo, path, last_commit)
+        last_metadata = metadata.Metadata.from_file(last_metadata_obj.data_stream)
+        last_param_metadata = last_metadata.flatten()[param_keys]
+        logging.debug(
+            f"Previous Metadata for {'/'.join(param_keys)}: {last_param_metadata}"
+        )
+        return last_param_metadata
+
+    async def get_previous_value(
+        self,
+        param_metadata: metadata.ParamMetadata,
+        param_keys: Tuple[str],
+        repo,
+        path: str,
+    ) -> Parameter:
+        """Get the last value for this parameter via git."""
+        logging.debug(f"Getting previous value for {'/'.join(param_keys)}")
+        prev_metadata = await self.get_previous_metadata(
+            param_metadata, param_keys, repo=repo, path=path
+        )
+        # TODO: get_update_serializer returns instantiated objects while the other
+        # getters return classes to be instantiated.
+        prev_serializer = params.get_update_serializer()
+        prev_update = get_update_handler(prev_metadata.theta_metadata.update_type)(
+            prev_serializer
+        )
+        return await prev_update.apply(prev_metadata, param_keys, repo=repo, path=path)
+
+    @abstractmethod
+    async def calculate_update(
+        self, parameter: Parameter, previous_parameter: Parameter
+    ) -> Parameter:
+        """Calculate the update required to go from previous_parameter -> parameter."""
+
+    @abstractmethod
+    async def apply_update(self, update: Parameter, previous: Parameter) -> Parameter:
+        """Apply the update to the previous value to get the new value."""
+
+    async def write_update(self, update: Parameter) -> metadata.LfsMetadata:
+        """Save and serialize (just) the update weights."""
+        serialized_update = await self.serializer.serialize({"parameter": update})
+        lfs_pointer = await git_utils.git_lfs_clean(serialized_update)
+        return metadata.LfsMetadata.from_pointer(lfs_pointer)
+
+    async def write(
+        self,
+        param: Parameter,
+        param_keys,
+        *,
```

Comment:
What's the purpose of the extra positional args here and in `apply`?

Reply:
My thought here is that some updates (non-incremental, i.e. dense) are somewhat distinct from incremental updates (i.e. sparse).

What this lets us do is have the shared parameters of the API (`param` and `param_keys` in this case) as the only thing that is "needed" in the non-incremental updates. This `*` actually forces arguments that follow to be passed by keyword. Then when calling code passes the extra parameters for the incremental updates they have to pass the extra things via keyword and those parameters are absorbed by the `**kwags` on the base update.

This creates a visual distinction for developers between parameters that all updates will need and ones that only some will need.

Another option would be to have conditionals in the calling code that decides which params to pass or to have the base class take all the parameters, even if they are unused (but that seemed to make it hard to infer what an update would do).

Reply:
Nice, this makes sense!